### PR TITLE
Add JUnit output tests and fix the issues.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Review by @sjrd and @gzm0.

All fixes were done after the creation of the testing tool, as such I saw them failing before I created the fixes.

File: junit-test/run-tests.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+jUnitTestJS = "jUnitTestJS"
+jUnitTestJVM = "jUnitTestJVM"
+
+cmd = ['sbt']
+
+def setFlagsInScope(flags, scope):
+    args = map(lambda x: '"-' + x + '"', flags)
+    cmd.append(
+        "set inScope(ThisScope in " + scope +")(jUnitTestArgs := Seq(" + \
+        ', '.join(args) + "))")
```
Comment (sjrd) [inline junit-test/run-tests.py]: `\` is not necessary inside the `()` of the call.
Also, consider using a `%`-formatted string:
```python
    cmd.append(
        "set inScope(ThisScope in %s)(jUnitTestArgs := Seq(%s))" %
        (scope, ", ".join(args)))
```

File: junit-test/run-tests.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+def testInSope(task, scope):
+    cmd.append(scope + "/" + task)
+
+for test in tests:
+
```
Comment (sjrd) [inline junit-test/run-tests.py]: Spurious blank line.

File: project/Build.scala [hunk 3]
```diff
@@ -1091,6 +1092,58 @@ object Build {
...
+      publishArtifact in Compile := false,
+      parallelExecution in Test := false,
+      unmanagedSourceDirectories in Test +=
+          baseDirectory.value.getParentFile / "shared/src/test/scala",
+
+      jUnitTestArgs := Seq("-v"),
+      testOptions += Tests.Argument(TestFrameworks.JUnit, jUnitTestArgs.value: _*),
+
+      checkJUnitTestFailuresOutput in Test := {
+        val _ = (testOnly in Test).toTask(" org.scalajs.junit.failure.*").failure.value
```
Comment (sjrd) [inline project/Build.scala]: Is there a way we could check that running *any* of the failure tests does indeed fail, rather than testing that when running *all* failure tests?

File: project/Build.scala [hunk 4]
```diff
@@ -1091,6 +1092,58 @@ object Build {
...
+      },
+
+      test in Test := {
+        val _ = (testOnly in Test).
+            toTask(" org.scalajs.junit.success.* org.scalajs.junit.namedecode.*").value
```
Comment (sjrd) [inline project/Build.scala]: That's a funky style. I suggest extracting the string in a temp val first:
```scala
val taskArgs = " org.scalajs.junit.success.* org.scalajs.junit.namedecode.*"
val _ = (testOnly in Test).toTask(taskArgs).value
```

File: project/JUnitTestLogger.scala [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        (level == Level.Error && msg.contains("Failed: Total"))
+      }
+
+      if (isTestRunFinished)
+        testRunFinished()
+      else if (areTestsFinished)
+        testsFinished()
+
+      log.append(logMsg)
+
```
Comment (sjrd) [inline project/JUnitTestLogger.scala]: Spurious blank line.

File: project/JUnitTestLogger.scala [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+          msg.replaceAll(", \\d+(\\.\\d*)?s", "")
+        else if (msg.matches(".*, took \\d+(\\.\\d*)? sec"))
+          msg.substring(0, msg.lastIndexOf(", took "))
+        else
+          msg
+      }
+      val logMsg = s"[$level] $msgWithoutTimes"
+
+      val isTestRunFinished = {
+        /* The first option is always printed at the start of a new test,
```
Comment (sjrd) [inline project/JUnitTestLogger.scala]: option -> alternative? message?

File: project/JUnitTestLogger.scala [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+      warnIgnored(referenceOutputIgnored, "reference")
+      warnIgnored(escapedColorLogIgnored, "actual")
+
+      val patch = DiffUtils.diff(referenceOutput2, escapedColorLog2)
+      val deltas = patch.getDeltas
+
+      if (deltas.isEmpty) {
+        loggerOutput(s"$SUCCESS JUnit output matched for $config")
+      } else {
+        loggerOutput(s"$ERROR JUnit output missmatch for $config")
```
Comment (sjrd) [inline project/JUnitTestLogger.scala]: typo: mismatch

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala [hunk 8]
```diff
@@ -261,6 +261,12 @@ object ScalaJSPlugin extends AutoPlugin {
         "scalaJSJavaSystemProperties",
         "List of arguments to pass to the Scala.js Java System.properties.",
         CTask)
+
+    val jUnitTestArgs = SettingKey[Seq[String]]("jUnitTestArgs",
+        "List of flags to pass to the JUnit framework in jUnitTest.", CSetting)
+
+    val checkJUnitTestFailuresOutput = TaskKey[Unit]("checkJUnitTestFailuresOutput",
+        "Executes tests in org.scalajs.junit.failure expecting them to fail.", DTask)
```
Comment (sjrd) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala]: Those should definitely not be in the published plugin. They are for our own build only. The should be in [`ExposedValues`](https://github.com/scala-js/scala-js/blob/v0.6.9/project/Build.scala#L45)

File: project/JUnitTestLogger.scala (hunk 6 above)
Comment (nicolasstucki) [inline project/JUnitTestLogger.scala]: Yes, should be message.

**Comment by sjrd** [comment]
In [the CI log](https://scala-webapps.epfl.ch/jenkins/job/scalajs-task-worker/114413/console), I see the following where the Python script should have run:
```
java.io.IOException: No locks available
	at sun.nio.ch.FileDispatcherImpl.lock0(Native Method)
	at sun.nio.ch.FileDispatcherImpl.lock(FileDispatcherImpl.java:91)
	at sun.nio.ch.FileChannelImpl.tryLock(FileChannelImpl.java:1095)
	at java.nio.channels.FileChannel.tryLock(FileChannel.java:1154)
	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:86)
	at xsbt.boot.Locks$GlobalLock.xsbt$boot$Locks$GlobalLock$$withChannelRetries$1(Locks.scala:78)
	at xsbt.boot.Locks$GlobalLock$$anonfun$withFileLock$1.apply(Locks.scala:97)
	at xsbt.boot.Using$.withResource(Using.scala:10)
	at xsbt.boot.Using$.apply(Using.scala:9)
	at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:58)
	at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:48)
	at xsbt.boot.Locks$.apply0(Locks.scala:31)
	at xsbt.boot.Locks$.apply(Locks.scala:28)
	at xsbt.boot.Launch.locked(Launch.scala:238)
	at xsbt.boot.Launch.app(Launch.scala:147)
	at xsbt.boot.Launch.app(Launch.scala:145)
	at xsbt.boot.Launch$.run(Launch.scala:102)
	at xsbt.boot.Launch$$anonfun$apply$1.apply(Launch.scala:35)
	at xsbt.boot.Launch$.launch(Launch.scala:117)
	at xsbt.boot.Launch$.apply(Launch.scala:18)
	at xsbt.boot.Boot$.runImpl(Boot.scala:41)
	at xsbt.boot.Boot$.main(Boot.scala:17)
	at xsbt.boot.Boot.main(Boot.scala)
Error during sbt execution: java.io.IOException: No locks available
```
but then it goes as if it succeeded. That does not look good.

File: project/Build.scala (hunk 4 above)
Comment (nicolasstucki) [inline project/Build.scala]: Oh, I can not do that. `toTask` macro expects its argument to be a `String` literal. 
```scala
[error] /Users/nastucki/GitHub/scala-js/project/Build.scala:1110: Illegal dynamic reference: taskArgs
[error]         val _ = (testOnly in Test).toTask(taskArgs).value
```

Comment (sjrd) [inline project/Build.scala]: Arf ... Well in that case the only reasonable way to adhere to the coding style is this:
```scala
val _ = {
  (testOnly in Test)
    .toTask(" org.scalajs.junit.success.* org.scalajs.junit.namedecode.*").value
}
```

**Comment by nicolasstucki** [comment]
Jenkins, retest this please.

**Comment by gzm0** [comment]
I have been thinking a lot about these tests and I really dislike that we need to introduce two new languages (as far as I am concerned, the mini-test language is a language as well) and another dependency on the CI servers (python) that is unmanaged.

Further, I am also not a fan, that test code and assertions are so far apart from each other.

Therefore I would like to propose another solution that uses JUnit to test JUnit.

```scala
import sbt.testing.{Logger, EventListener, TaskDef, Tasks}

abstract class JUnitTest {
  val jUnitFramework = new com.novocode.junit.JUnitFramework()

  val suiteUnderTestName = getClass.getName.stripSuffix("Assertions")

  // records everything including invocation order and result.
  val recorder: Logger with EventListener = ???

  // appropriate class loader for platform, needs platform extension
  val classLoader = ???
  val classLoaderJVM = getClass.getClassLoader
  val classLoaderJS =
    new org.scalajs.testinterface.ScalaJSClassLoader(js.Dynamic.global)

  // To be overriden by subclasses
  def frameworkArgs(): Array[String] = Array.empty

  // run all tasks and the tasks they generate, needs platform extension
  private def executeLoop(tasks: Array[Task]): Unit = ???

  private def executeLoopJVM(): Unit = {
    if (tasks.nonEmpty) {
      executeLoop(tasks.flatMap(_.execute(recorder, Array(recorder))))
    }
  }

  private def executeLoopJS(): Unit = {
    if (tasks.nonEmpty) {
      executeLoop(tasks.flatMap { task => 
        /* This is a hack, assuming the asynchronous continuation will be called before
         * the outer call returns. This is terrible in general, but we do not have
         * another choice, since JUnit does not support async testing.
         * However, since JUnit does not support async testing, this is a safe 
         * assumption to make (module JUnit implementation details, which we control).
         * If JUnit ever were to support async testing, we could also change this here.
         */
        var newTasks: Array[Task] = null
        task.execute(recorder, Array(recorder), newTasks = _)
        newTasks
      })
    }
  }

  @Before
  def setUp(): Unit = {
    val runner = jUnitFramework.runner(frameworkArgs, Array.empty, classLoader)
    val tasks = runner.tasks(new TaskDef(suiteUnderTestName,
        jUnitFramework.fingerprints.head, explicitlySpecified = true,
        selectors = Array.empty))
    executeLoop(tasks)
    recorder.recordDone(runner.done())
  }
}
```

Now, a test could read:

```scala
class AssertTrueTest {
  @Test def test(): Unit = {
    assertTrue(false)
  }
}

class AssertTrueTestAssertions {
  @Test def recordsFailure(): Unit = {
    assertEquals(recorder.failureCount, 1)
  }
}
```

The only thing left to do is to tell sbt not to invoke tests that do not end in "Assertions":

```scala
testOptions in Test += Seq(Tests.Filter(_.endsWith("Assertions")))
```

IMHO, this will lead to much more readable and maintainable tests. Furthermore, we also do not rely on sbt's behavior in our tests but can test against the true interface of the testing framework itself.

WDYT?

**Comment by sjrd** [comment]
Makes sense. It's definitely worth a try.

**Comment by nicolasstucki** [comment]
I will try this approach.
PS The current approach with the sbt extra seams fail most of the times on Jenkins. I have not been able to make it work in a stable way on that platform.

**Comment by nicolasstucki** [comment]
@sjrd @gzm0 here is a version based on @gzm0 idea. It is clearly the correct way to achieve these tests. 

The test are checking the same characteristic of the output as before, plus the exact moment when events are triggered.

**Comment by nicolasstucki** [comment]
Jenkins, retest this please.

File: junit-test/output-js/src/main/scala/org/scalajs/junit/utils/JUnitTestPlatformImpl.scala [hunk 9]
```diff
@@ -0,0 +1,27 @@
+package org.scalajs.junit.utils
+
+import sbt.testing._
+
+import scala.scalajs.js
+
+object JUnitTestPlatformImpl {
+
+  def getClassLoader: ClassLoader =
+      new org.scalajs.testinterface.ScalaJSClassLoader(js.Dynamic.global)
```
Comment (sjrd) [inline junit-test/output-js/src/main/scala/org/scalajs/junit/utils/JUnitTestPlatformImpl.scala]: 2 spaces

File: junit-test/output-jvm/src/main/scala/org/scalajs/junit/JUnitTestLoggerOutputCheck.scala [hunk 10]
```diff
@@ -0,0 +1,33 @@
+package org.scalajs.junit
+
+import java.io.File
+
+object JUnitTestLoggerOutputCheck extends App {
```
Comment (sjrd) [inline junit-test/output-jvm/src/main/scala/org/scalajs/junit/JUnitTestLoggerOutputCheck.scala]: This file seems to be dead code.

File: junit-test/shared/src/test/scala/org/scalajs/junit/AssertEquals2Test.scala [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+class AssertEquals2TestAssertions extends JUnitTest with FailureFrameworkArgs {
+
+  override val expectedFail: Int = 1
+  override val expectedTotal: Int = 1
+
+  protected def expectedOutput(context: OutputContext): List[Output] = {
+    import context._
+    List(
+      testRunStartedOutput,
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/AssertEquals2Test.scala]: 4 spaces.
This is probably the same in all `expectedOutput` methods.

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/FailureFrameworkArgs.scala [hunk 12]
```diff
@@ -0,0 +1,19 @@
+package org.scalajs.junit.utils
+
+trait FailureFrameworkArgs {
+  protected def frameworkArgss: List[List[String]] = List(
+    Nil,
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/FailureFrameworkArgs.scala]: 4 spaces here too

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/SuccessFrameworkArgs.scala [hunk 13]
```diff
@@ -0,0 +1,12 @@
+package org.scalajs.junit.utils
+
+trait SuccessFrameworkArgs {
+  protected def frameworkArgss: List[List[String]] = List(
+    Nil,
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/SuccessFrameworkArgs.scala]: 4 spaces here too.

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+      if (noColor) str
+      else MAGENTA + str + NORMAL
+
+    private def cyan(str: String): String =
+      if (noColor) str
+      else CYAN + str + NORMAL
+
+  }
+
+  protected class JunitTestRecorder(context: OutputContext)
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: Uppercase `U` in `JUnit`

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+  }
+
+  protected sealed trait Output
+  protected case class Info(msg: String) extends Output
+  protected case class Warn(msg: String) extends Output
+  protected case class Error(msg: String) extends Output
+  protected case class Debug(msg: String) extends Output
+  protected case class Trace(msg: String) extends Output
+  protected case class Event(status: String) extends Output
+  protected case class Done(msg: String) extends Output
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: case classes should be `final`

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+    def formattedTestClass: String = {
+      val (pack, cls) =
+        suiteUnderTestName.splitAt(suiteUnderTestName.lastIndexOf('.'))
+      pack + '.' + yellow(cls.tail)
+    }
+
+    def formattedAssumptionViolatedException: String =
+      "org.junit.internal." + red("AssumptionViolatedException")
+
+
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: Double blank line.

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+    def trace(t: Throwable): Unit = buff += Trace(t.getMessage)
+
+    def handle(event: sbt.testing.Event): Unit = {
+      buff += Event(event.status().toString)
+    }
+
+    def recordDone(msg: String): Unit = {
+      buff += Done(msg)
+      checkOutput(buff.result())
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: IMO `checkOutput` should be called explicitly from `testJUnitOutput`, rather than implicitly through `recordDone`.

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+      val enc1 =
+        msg.replaceFirst(", took \\d+(\\.\\d*)? sec", s", took $TIME_TAG sec")
+      if (!enc1.contains("Test run finished")) enc1
+      else enc1.replaceFirst("\\d+(\\.\\d*)?s", TIME_TAG + "s")
+    }
+
+    private def checkOutput(actual: List[Output]): Unit = {
+      val expected = expectedOutput(context)
+      type ListTuple = (List[Output], List[Output])
+      def minimizeDiff(lists: ListTuple): ListTuple = {
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: Blank line above this line.

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+        lists match {
+          case (x::xs, y::ys) if x == y => minimizeDiff(xs, ys)
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: Spaces around `::`

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+    private def encodeTimes(msg: String): String = {
+      val enc1 =
+        msg.replaceFirst(", took \\d+(\\.\\d*)? sec", s", took $TIME_TAG sec")
+      if (!enc1.contains("Test run finished")) enc1
+      else enc1.replaceFirst("\\d+(\\.\\d*)?s", TIME_TAG + "s")
+    }
+
+    private def checkOutput(actual: List[Output]): Unit = {
+      val expected = expectedOutput(context)
+      type ListTuple = (List[Output], List[Output])
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: Blank space above this line.

Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: Or move `val expected` right before `val (ex, ac)`

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala (hunk 19 above)
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: Also don't rely on auto-tupling. Add `()` around `xs, ys`.
Alternatively, have `minimizeDiff` take two parameters of type `List[Output]` rather than a `ListTuple`.

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala (hunk 18 above)
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: This method should be annotated with `@tailrec`, I believe.

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+      def minimizeDiff(lists: ListTuple): ListTuple = {
+        lists match {
+          case (x::xs, y::ys) if x == y => minimizeDiff(xs, ys)
+          case _                        => lists
+        }
+      }
+
+      val (ex, ac) = minimizeDiff((expected, actual))
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: Shouldn't you also try and minimize the diff from the end? For example add
```scala
val (ex2, ac2) = minimizeDiff((ex.reverse, ac.reverse))
val ex3 = ex2.reverse
val ac3 = ac3.reverse
```

File: project/Build.scala [hunk 22]
```diff
@@ -461,6 +461,7 @@ object Build {
               clean in stubs, clean in cli,
               clean in testInterface, clean in jasmineTestFramework,
               clean in jUnitRuntime, clean in jUnitPlugin,
+              clean in jUnitTestOutputsJS, clean in jUnitTestOutputsJS,
```
Comment (sjrd) [inline project/Build.scala]: JS twice. One of them should be JVM

File: project/Build.scala [hunk 23]
```diff
@@ -1092,6 +1093,40 @@ object Build {
       fatalWarningsSettings ++ Seq(name := "Scala.js JUnit test runtime")
   ).withScalaJSCompiler.dependsOn(testInterface)
 
+  val commonJUnitTestOutputsSettings = commonSettings ++ fatalWarningsSettings ++ Seq(
+      publishArtifact in Compile := false,
+      parallelExecution in Test := false,
+      unmanagedSourceDirectories in Test +=
+          baseDirectory.value.getParentFile / "shared/src/test/scala",
```
Comment (sjrd) [inline project/Build.scala]: 2 spaces

File: project/Build.scala [hunk 24]
```diff
@@ -1092,6 +1093,40 @@ object Build {
...
+      testOptions += Tests.Argument(TestFrameworks.JUnit, "-v", "-a", "-s"),
+      test in Test := {
+        val _ = {
+          (testOnly in Test).toTask(" *Assertions").result.value
+        }
```
Comment (sjrd) [inline project/Build.scala]: Doesn't the `val _ = ...` fit on one line?

File: project/Build.scala [hunk 25]
```diff
@@ -1092,6 +1093,40 @@ object Build {
...
+      }
+  )
+
+  lazy val jUnitTestOutputsJS = Project(
+    id = "jUnitTestOutputsJS",
```
Comment (sjrd) [inline project/Build.scala]: 4 spaces

File: project/Build.scala [hunk 26]
```diff
@@ -1092,6 +1093,40 @@ object Build {
...
+    base = file("junit-test/output-js"),
+    settings = commonJUnitTestOutputsSettings ++ myScalaJSSettings ++ Seq(
+      name := "Tests for Scala.js JUnit output in JS."
+    )
+  ).withScalaJSCompiler.withScalaJSJUnitPlugin.dependsOn(jUnitRuntime, testInterface)
+
+
+  lazy val jUnitTestOutputsJVM = Project(
+    id = "jUnitTestOutputsJVM",
```
Comment (sjrd) [inline project/Build.scala]: 4 spaces

File: project/Build.scala [hunk 27]
```diff
@@ -1092,6 +1093,40 @@ object Build {
...
+    base = file("junit-test/output-jvm"),
+    settings = commonJUnitTestOutputsSettings ++ Seq(
+      name := "Tests for Scala.js JUnit output in JVM.",
+      libraryDependencies ++= Seq(
+          "org.scala-sbt" % "test-interface" % "1.0",
```
Comment (sjrd) [inline project/Build.scala]: `% "test"`?

File: project/Build.scala [hunk 28]
```diff
@@ -1092,6 +1093,40 @@ object Build {
...
+      }
+  )
+
+  lazy val jUnitTestOutputsJS = Project(
+    id = "jUnitTestOutputsJS",
+    base = file("junit-test/output-js"),
+    settings = commonJUnitTestOutputsSettings ++ myScalaJSSettings ++ Seq(
+      name := "Tests for Scala.js JUnit output in JS."
+    )
+  ).withScalaJSCompiler.withScalaJSJUnitPlugin.dependsOn(jUnitRuntime, testInterface)
```
Comment (sjrd) [inline project/Build.scala]: `% "test"` for both dependencies?

File: project/Build.scala (hunk 24 above)
Comment (sjrd) [inline project/Build.scala]: And, actually, why not manipulating `testFilter` instead, as suggested by @gzm0?

File: junit-test/shared/src/test/scala/org/scalajs/junit/MethodNameDecodeTest.scala [hunk 29]
```diff
@@ -0,0 +1,35 @@
+package org.scalajs.junit
+
+import org.junit.Test
+import org.scalajs.junit.utils.JUnitTest
+
+class MethodNameDecodeTest {
+  @Test def `abcd ∆ƒ 😀 * #&$`(): Unit = ()
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/MethodNameDecodeTest.scala]: I actually recommend using `\uD83D\uDE00` instead of the actual astral character. IDEs don't like astral characters. Since `\u` processing happens "pre-lexer" in Scala, it's fine to do so.

File: junit-test/shared/src/test/scala/org/scalajs/junit/MultiAssumeFail1Test.scala [hunk 30]
```diff
@@ -0,0 +1,46 @@
...
+import org.junit.Assume._
+import org.junit.Test
+
+import org.scalajs.junit.utils.{JUnitTest, SuccessFrameworkArgs}
+
+class MultiAssumeFail1Test {
+  @Test def multiTest1(): Unit = {
+    assumeTrue("This assume should not pass", false)
+  }
+  @Test def multiTest2(): Unit = ()
```
Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/MultiAssumeFail1Test.scala]: Blank line above this line.

**Comment by sjrd** [comment]
That's all I have for the last commit.

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala (hunk 15 above)
Comment (nicolasstucki) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: If I I get a compile error saying `The outer reference in this type test cannot be checked at run time.`. But I can use `private final` as they also never escape this class.

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala (hunk 21 above)
Comment (nicolasstucki) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: Yes, but then the diff becomes a useless because it is impossible to know where it should be in the list. I solved this problem by printing in grey all the elements that were dropped by `minimizeDiff`. 

The output will be like:
screen shot 2016-05-20 at 11 56 44

Where the diff is clearly the additional `Event(Failure)`. I'll also remove the color from the non diff lines.

Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: Maybe you could keep 3 lines above and below the differences? (like the default behavior of `diff -u`)

**Comment by nicolasstucki** [comment]
This PR might be ready to merge.

File: junit-runtime/src/main/scala/org/scalajs/junit/JUnitExecuteTest.scala [hunk 31]
```diff
@@ -64,147 +73,145 @@ final class JUnitExecuteTest(taskDef: TaskDef, runner: JUnitBaseRunner,
...
+        logFormattedDebug(decodedMethodName,
+            s"finished, took ${getTimeInSeconds()} sec")
+
+        if (expectedException == classOf[org.junit.Test.None]) {
+          testPassed(methodName)
+          executeAfterMethods()
+        } else {
+          val msg = {
+            s"failed: Expected exception: " + expectedException +
+                s"took ${getTimeInSeconds()} sec"
```
Comment (sjrd) [inline junit-runtime/src/main/scala/org/scalajs/junit/JUnitExecuteTest.scala]: 0 space.

File: junit-runtime/src/main/scala/org/scalajs/junit/JUnitExecuteTest.scala [hunk 32]
```diff
@@ -64,147 +73,145 @@ final class JUnitExecuteTest(taskDef: TaskDef, runner: JUnitBaseRunner,
...
-          true
-
-        case ex: internal.AssumptionViolatedException =>
-          logFormattedInfo(method.name, "started")
-          logAssertionWarning(method.name, ex, timeInSeconds())
-          true
+        case ex: Throwable =>
+          val timeInSeconds = getTimeInSeconds()
+          if (classOf[AssumptionViolatedException].isInstance(ex) ||
+              classOf[internal.AssumptionViolatedException].isInstance(ex)) {
```
Comment (sjrd) [inline junit-runtime/src/main/scala/org/scalajs/junit/JUnitExecuteTest.scala]: Why not `ex.isInstanceOf[AssumptionViolatedException] || ex.isInstanceOf[internal.AssumptionViolatedException`?

File: junit-runtime/src/main/scala/org/scalajs/junit/JUnitExecuteTest.scala [hunk 33]
```diff
@@ -64,147 +73,145 @@ final class JUnitExecuteTest(taskDef: TaskDef, runner: JUnitBaseRunner,
...
+            logAssertionWarning(decodedMethodName, ex, timeInSeconds)
+            testSkipped()
+          } else if (expectedException.isInstance(ex)) {
+            testPassed(methodName)
+            executeAfterMethods()
+          } else if (expectedException == classOf[org.junit.Test.None]) {
+            val isAssertion = ex.isInstanceOf[AssertionError]
+            val failedMsg = new StringBuilder
```
Comment (sjrd) [inline junit-runtime/src/main/scala/org/scalajs/junit/JUnitExecuteTest.scala]: Consider a `var failedMsg = ""`

Comment (sjrd) [inline junit-runtime/src/main/scala/org/scalajs/junit/JUnitExecuteTest.scala]: Ah nvm, that was like that before. Let's not change it in an unrelated commit.

File: ci/matrix.xml [hunk 34]
```diff
@@ -78,6 +78,9 @@
 
   <task id="test-suite-ecma-script5"><![CDATA[
     setJavaVersion $java
+    sbtretry ++$scala jUnitTestOutputsJVM/test jUnitTestOutputsJS/test \
+        'set scalaJSUseRhino in Global := false' jUnitTestOutputsJS/test \
+        'set scalaJSStage in Global := FullOptStage' jUnitTestOutputsJS/test &
```
Comment (sjrd) [inline ci/matrix.xml]: Shouldn't this be `&&`?

File: ci/matrix.xml [hunk 35]
```diff
@@ -143,6 +146,9 @@
 
   <task id="test-suite-ecma-script6"><![CDATA[
     setJavaVersion $java
+    sbtretry ++$scala jUnitTestOutputsJVM/test jUnitTestOutputsJS/test \
+        'set scalaJSUseRhino in Global := false' jUnitTestOutputsJS/test \
+        'set scalaJSStage in Global := FullOptStage' jUnitTestOutputsJS/test &
```
Comment (sjrd) [inline ci/matrix.xml]: Shouldn't this be `&&`?

**Comment by sjrd** [comment]
That's all.

File: junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala [hunk 36]
```diff
@@ -0,0 +1,251 @@
...
+  }
+
+  protected sealed trait Output
+  private final case class Info(msg: String) extends Output
+  private final case class Warn(msg: String) extends Output
+  private final case class Error(msg: String) extends Output
+  private final case class Debug(msg: String) extends Output
+  private final case class Trace(msg: String) extends Output
+  private final case class Event(status: String) extends Output
+  private final case class Done(msg: String) extends Output
```
Comment (nicolasstucki) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: On `2.12.0-M2` we get the following issue since I added back the `final`, I will have to remove the `final`.
```scala
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:60: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Info(msg: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:61: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Warn(msg: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:62: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Error(msg: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:63: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Debug(msg: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:64: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Trace(msg: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:65: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Event(status: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:66: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Done(msg: String) extends Output
[error]                            ^
```

Comment (sjrd) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: I suggest you put them in the companion object. Together with `Output` as well (which must then be public). You can add a type alias in the class for `Output`: `type Output = JUnitTest.Output` for convenience in subclasses.

case classes nested in classes and traits always have tricky behaviors.

Comment (nicolasstucki) [inline junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala]: That solution seams to work.

**Comment by sjrd** [comment]
LGTM
